Vg trace

taoshi-117 opened the pull request:
add vg trace in the /pyCalcilations/field_tracer.py

taoshi-117 committed:
added vg trace in fieldtracer.py

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -190,89 +190,140 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
-      points_traced.append( list(points) )             # list of lists of 3-element arrays
-
-   return points_traced
+   if fg is not None:
```
Here I would take the fg routine and place it into a separate function

taoshi-117 replied:
So I should define the function earlier and call it here?

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -190,89 +190,140 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+         points_traced.append( list(points) )             # list of lists of 3-element arrays
+   
+   elif vg is not None:
+      # Search for the unique coordinates in the given seeds only
```
... and here I would take the vg routine and place that in a separate function

alhom replied:
I mean, these functions I would place outside the static_field_tracer function entirely.

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -190,89 +190,140 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+      unique_seed_coords,indices = np.unique(seed_coords, axis = 0, return_inverse = True)    # indice here is to reverse the coords order to initial
+      n_unique_seeds = unique_seed_coords.shape[0]
+      points_traced_unique = np.zeros((n_unique_seeds, max_iterations, 3))
+      print(vg)
```
No need for debug printout anymore here

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -190,89 +190,140 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+      if grid_var.startswith("fg"):
+         fg = grid_var
+
+      elif grid_var.startswith("vg"):   
```
Hm. We may actually want to trace variables with names like "proton/vg_v". string,split('/') and check if any of the substrings starts with vg?

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -190,89 +190,140 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+         points_traced_unique[mask_update,i,:] = next_points[mask_update,:]
+
+         distances = np.linalg.norm(points_traced_unique[:,i,:],axis = 1)
+         mask_update[(distances <= 5*Re) | (distances >= 40*Re)] = False
```
`(distances <= 5*Re) | (distances >= 40*Re)` should rather be a function that can be passed to the tracer.

So that the function passed accepts the array of points and tells the tracer that these traces can be stopped.

taoshi-117 committed:
Update fieldtracer.py

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -190,89 +190,156 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+         return val_at_point/val_mag
+
+      # The stop condition for tracing, unit: m
+      def stopping_condition(distances, lower_bound, upper_bound):
```
This I would generalize and remove entirely from this scope, so that the function
`def static_field_tracer_3d( vlsvReader, seed_coords, max_iterations, dx, direction='+', grid_var = 'vg_b_vol', lower_bound = 5 * 6371000, upper_bound = 40 * 6371000):`
would rather be
`def static_field_tracer_3d( vlsvReader, seed_coords, max_iterations, dx, direction='+', grid_var = 'vg_b_vol', stopping_condition = default_stopping_condition):`

with a new default function that returns False for each point. Then you do in your own script your own function like:

```
lower_bound = 30e6 # or whatever
upper_bound = 1e9
def my_stop(points):
   distances = np.linalg.norm(points[:,:],axis = 1)
   return (distances <= lower_bound) | (distances >= upper_bound)
```
...

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -190,89 +190,156 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+         points_traced_unique[mask_update,i,:] = next_points[mask_update,:]
+
+         distances = np.linalg.norm(points_traced_unique[:,i,:],axis = 1)
+         mask_update[stopping_condition(distnaces, lower_bound, upper_bound)] = False
```
and here the mask update would then rather be
```
mask_update[stopping_condition(points_traced_unique[:,i,:])] = False
```

alhom replied:
And there's a typo: distnaces..

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -190,89 +190,156 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+      if "fg" in grid_var:
+         fg = grid_var
+
+      elif "vg" in grid_var:   
```
a bit risky, in case we would have variable names with fg/vg in them elsewhere than as a prefix (not that I think we have/should have any). See the previous note on string.split('/') though, since then we can check again for a prefix.

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -190,89 +190,156 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+
+      elif "vg" in grid_var:   
+         vg = grid_var   
+         neighbors_vg = vlsvReader.read_variable_to_cache("vg_regular_interp_neighbors")
```
Let's have this neighbors caching call outside of the whole script (so the user can call this before calling static_field_tracer_3d if the want to) - I need to rethink this caching anyway.

taoshi-117 committed:
Update2 fieldtracer.py

taoshi-117 committed:
Update2 fieldtracer.py

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -273,6 +219,129 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+      forward = static_field_tracer_3d(vlsvReader, coord_list, max_iterations, dx, direction='+', grid_var = grid_var)
+      return np.concatenate((backward,forward[:, 1:, :]), axis = 1)
+
+   multiplier = -1 if direction == '-' else 1   
```
I don't think this is passed to the tracing subroutines now.

taoshi-117 replied:
Yes, and neither does dx

taoshi-117 committed:
Update2_1 fieldtracer.py

alhom commented on pyCalculations/fieldtracer.py:
```diff
@@ -223,7 +223,7 @@ def fg_trace(fg, seed_coords, max_iterations):
 
 # vg tracing for static_field_tracer_3d
-def vg_trace(vg, seed_coords, max_iterations, stop_condition):
+def vg_trace(vg, seed_coords, max_iterations, dx, multiplier, stop_condition):
```
This still needs to have vlsvReader passed in!

taoshi-117 committed:
Update3 fieldtracer.py

1. passed in vlsvReader
2. added stop_condition in fg trace, and vectorised a bit

taoshi-117 committed:
Update fieldtracer.py

fixed the Nameerror

taoshi-117 committed:
Update calculations.py

add import static_field_tracer_3d
